include/revng/ADT/UpcastablePointer.h

```diff
@@ -174,6 +174,10 @@ class UpcastablePointer {
     return UpcastablePointer<T>(new Q(std::forward<Args>(TheArgs)...));
   }
 
+  UpcastablePointer copy() const {
```

Comment:
Why did you choose the name  `copy` when there's `clone` right below?

Reply:
To differentiate the two actually. But if you'd prefer them to have the same name, that works just as well.

Reply:
Maybe `clone` and `cloneInternal`?  Or `copy` and `copyInternal`?

But looking more deeply... The current copy constructor seems utterly broken. `UpcasteblePointer` should have "copy semantics", similarly to `std::variant` and the current copy constructor duplicates the inner pointer!
Am I missing something or should we simply drop `copy` and implement a correct copy constructor?

Reply:
1. Wow, how did I not notice that?! I guess because `operator=` right before it is fine. But more importantly, how did it work?

2. I'd still prefer to have an explicit `copy/clone`. To be honest, I'd rather we could avoid having the constructor, but TTG demands a sacrifice /s

Reply:
I understand your desire for being explicit, but the idea of `UpcastablePointer` is to have (almost) the same properties of `std::variant` (even if we won't be able to ever use `std::variant` due to loops).

In fact, `UpcastablePointer` is a bad name and we should (not now) rename it to something that recalls `variant`.

Reply:
1. I don't know, to me it doesn't feel that variant-like. Especially considering that the kind "label" is inside the objects. It's a lot close to the polymorphic relation (it, by definition), is.

> In fact, `UpcastablePointer` is a bad name

2. Actually, I like it quite a bit - it explains exactly what it is and what it does.

Reply:
1. I don't know, to me it doesn't feel that variant-like. Especially considering that the kind "label" is inside the objects. It's a lot close to the polymorphic relation (it, by definition), is.

> In fact, `UpcastablePointer` is a bad name

2. Actually, I like it quite a bit - it explains exactly what it is and what it does.

Reply:
I have second thoughts on this. Can we at least keep it (with whichever name you want, it doesn't really matter) until the next `clang` update, where we'll get `auto{}` for expressing this kind of logic?

Consider that it makes it a lot easier to use move-first interfaces:
```cpp
void do_stuff(model::UpcastableType &&);

// I have to give my type away:
do_stuff(std::move(MyType));

// And I do not:
do_stuff(MyType.copy());
```

Reply:
> we'll get auto{} for expressing this kind of logic

Uh?

> Consider that it makes it a lot easier to use move-first interfaces:

This argument could be done for any class anywhere, I don't see why `UpcastableType` should be special.
If we don't have a copy constructor/assignment we're disabling the copy constructor/assignment in any container containing this (e.g., `vector<UpcastableType<...>>`).

`UpcastablePointer` should be like `std::variant`, but with support for recursive references.

Reply:
> Uh?

Explicit copy.

> This argument could be done for any class anywhere, I don't see why `UpcastableType` should be special.

It's special in the sense that its name is really long and it's used a lot, leading us to either force `using UT = UpcastableType` everywhere or providing some custom helpers, at least until `auto{}` arrives.

Reply:
1. I've no idea what you mean when you talk about `auto{}`. `auto{}` where?
2. I don't understand... What does the length of the name have to do with what we're talking about? We're talking about accidental copies and "move-first interface".
3. It seems quite reasonable to me.

I don't see why we should prevent a `vector<UpcastablePointer<...>>` from being copiable. `unique_ptr` has "reference semantics" so it make sense that it cannot be copied trivially, `UpcastablePointer` was designed to have value semantics, so it should be copiable.

Reply:
1. As in, a short way to show an explicit copy. It's a `c++23` feature, so after the next clang update, we'll be able to just say `auto(MyThing)` instead of `MyThingsType(MyThing)`.

2. We cannot forbid accidental copies. Because sorted vectors require those. I wish we could, but that's way too much work for now. This is all about making `move-first` interfaces easier to work with.

3. In theory, yes, in practice it means we add `using UT = UpcastableType` everywhere.

4. We are not preventing anything from being copied. I wanted to do that, but trying to pitch you the `UpcastableReference` idea, but we decided on "maybe one day", which is reasonable. The `copy` thing is just an explicit alternative to a copy constructor (which by chance also makes it easier to use move-first interfaces), not a replacement of it.

Reply:
So you want to keep the copy constructor?

> I'd still prefer to have an explicit copy/clone. 

I guess I misunderstood this sentence then and kinda ignored the next one.

If you want to have `.copy` *on top* of the copy constructor, OK.

Reply:
> So you want to keep the copy constructor?

Getting rid of it requires way to much effort, we need to get this in already, it's getting way too big.

> If you want to have `.copy` on top of the copy constructor, OK.

Yes!